### PR TITLE
New Feature: Stream from Page

### PR DESCRIPTION
**Pull request by bigdale123**
So, after I got page streaming working in https://github.com/koreader/koreader/pull/9650 I noticed that every time you open a book, it started at the beginning. I wanted to add progress pulling, but progress is served on different endpoints on different servers.

Stream from Page allows you to enter a page number and the Image Viewer jumps to that page, rather than having to turn through all the pages prior to where you left off.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9664)

### CODE REVIEW

**Comment by bigdale123** [comment]
I got too excited and forgot to remove a testing variable...

**Comment by bigdale123** [comment]
Also curious, is "Stream from page" misleading? Should it be renamed?

File: plugins/opds.koplugin/opdsbrowser.lua [hunk 1]
```diff
@@ -695,7 +695,40 @@ function OPDSBrowser:downloadFile(item, filename, remote_url)
...
 
-function OPDSBrowser:streamPages(item, remote_url, count)
+-- This function shows a dialog with input fields
+-- for entering information for Page streaming from.
+-- an input page number
+function OPDSBrowser:jumpToPage(viewer)
+    local input_dialog
+    input_dialog = InputDialog:new{
+        title = _("Enter Page Number"),
+        input = "",
```
Comment (hius07) [inline plugins/opds.koplugin/opdsbrowser.lua:705]: The line can be removed

File: plugins/opds.koplugin/opdsbrowser.lua [hunk 2]
```diff
@@ -695,7 +695,40 @@ function OPDSBrowser:downloadFile(item, filename, remote_url)
...
+        input_hint = "Page #",
```
Comment (hius07) [inline plugins/opds.koplugin/opdsbrowser.lua:706]: Do you know the maximum page number? It can be shown in the hint, eg (1 - 250).

File: plugins/opds.koplugin/opdsbrowser.lua [hunk 3]
```diff
@@ -695,7 +695,40 @@ function OPDSBrowser:downloadFile(item, filename, remote_url)
...
+                    callback = function()
+                        UIManager:close(input_dialog)
+                    end,
+                },
+                {
+                    text = _("Stream"),
+                    is_enter_default = true,
+                    callback = function()
+                        UIManager:close(input_dialog)
+                        viewer:switchToImageNum(tonumber(input_dialog:getInputValue()))
```
Comment (hius07) [inline plugins/opds.koplugin/opdsbrowser.lua:721]: You can add `input_type == "number"`, `getInputValue` makes `tonumber` for you.

**Comment by hius07** [comment]
Can you show the sreenshot of the download dialog please.

File: plugins/opds.koplugin/opdsbrowser.lua [hunk 4]
```diff
@@ -695,7 +695,40 @@ function OPDSBrowser:downloadFile(item, filename, remote_url)
...
 end
 
-function OPDSBrowser:streamPages(item, remote_url, count)
+-- This function shows a dialog with input fields
+-- for entering information for Page streaming from.
+-- an input page number
+function OPDSBrowser:jumpToPage(viewer)
+    local input_dialog
+    input_dialog = InputDialog:new{
+        title = _("Enter Page Number"),
```
Comment (hius07) [inline plugins/opds.koplugin/opdsbrowser.lua:704]: "Enter page number"

File: plugins/opds.koplugin/opdsbrowser.lua (hunk 3 above)
Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua:721]: > This will also switch the keyboard to the symbols/numbers layer.

I Did not know this, I will change it in the next commit!

**Comment by bigdale123** [comment]
> Can you show the sreenshot of the download dialog please.
Here's two, one with a download button and another without.
image
image

File: plugins/opds.koplugin/opdsbrowser.lua (hunk 3 above)
Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua:721]: Leaving a note for myself, this is done.

**Comment by hius07** [comment]
> Here's two, one with a download button and another without.

Your two stream buttons should form a row, and be added below download buttons.

Maybe this dialog should be redesigned when there are no download buttons.

File: plugins/opds.koplugin/opdsbrowser.lua (hunk 1 above)
Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua:705]: Leaving a note for myself, this is done.

File: plugins/opds.koplugin/opdsbrowser.lua (hunk 2 above)
Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua:706]: Leaving a note for myself, this is done.

File: plugins/opds.koplugin/opdsbrowser.lua (hunk 4 above)
Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua:704]: Leaving a note for myself, this is done.

**Comment by bigdale123** [comment]
> > Here's two, one with a download button and another without.
> 
> Your two stream buttons should form a row, and be added below download buttons.
> 
> Maybe this dialog should be redesigned when there are no download buttons.

What if I put a separator between the download button and page stream buttons? That would force it to be two rows.

**Comment by bigdale123** [comment]
Like This?
image
image

**Comment by poire-z** [comment]
Do filename appear always like this?:
image
These probably should be url-decoded (https://www.google.com/search?hl=en&num=25&q=lua+url+decode), for display and for the filename they get saved to.

**Comment by bigdale123** [comment]
> Do filename appear always like this?: image These probably should be url-decoded (https://www.google.com/search?hl=en&num=25&q=lua+url+decode), for display and for the filename they get saved to.

Let me try it out with other files on my server, if it looks like garbage for all of them i'll try that.

**Comment by bigdale123** [comment]
I didn't want to include another library (partially because I didn't know how I would bundle it in with KOReader :) ), so I just found some open source code that does the trick and cited it.

**Comment by bigdale123** [comment]
Screenshot
image

**Comment by poire-z** [comment]
> I didn't want to include another library [...] so I just found some open source code that does the trick and cited it

That's just fine for a tiny task like this :)

File: plugins/opds.koplugin/opdsbrowser.lua [hunk 5]
```diff
@@ -695,7 +695,47 @@ function OPDSBrowser:downloadFile(item, filename, remote_url)
...
+                    callback = function()
+                        UIManager:close(input_dialog)
+                    end,
+                },
+                {
+                    text = _("Stream"),
+                    is_enter_default = true,
+                    callback = function()
+                        UIManager:close(input_dialog)
+                        if (input_dialog:getInputValue() > count) then
```
Comment (hius07) [inline plugins/opds.koplugin/opdsbrowser.lua:722]: You will get a crash on entering a string that cannot be converted to number.

File: plugins/opds.koplugin/opdsbrowser.lua [hunk 6]
```diff
@@ -804,11 +859,20 @@ function OPDSBrowser:showDownloads(item)
...
+    -- handles adding download button(s)
+    if (download_buttons ~= {}) then
+        if (#download_buttons == 1) then
+            table.insert(buttons, download_buttons)
+            table.insert(buttons, {})  -- add separator for stream buttons
+        elseif (#download_buttons % 2 == 1) then
+            table.insert(download_buttons, {text = ""})
+        end
+    end
+    if (#stream_buttons % 2 == 1) then -- we need even number of type buttons
```
Comment (hius07) [inline plugins/opds.koplugin/opdsbrowser.lua:871]: Don't you always have pairs of the stream buttons?

File: plugins/opds.koplugin/opdsbrowser.lua (hunk 5 above)
Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua]: Added a check to make sure input is not nil. If it is nil, an error message pops up.

File: plugins/opds.koplugin/opdsbrowser.lua (hunk 6 above)
Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua]: Now that I think about it, that's true. I'll remove that.

File: plugins/opds.koplugin/opdsbrowser.lua (hunk 5 above)
Comment (hius07) [reply plugins/opds.koplugin/opdsbrowser.lua]: We usually just do not accept incorrect input, so it can be smth like
```Lua
local page_num = input_dialog:getInputValue()
if page_num then
    UIManager:close(input_dialog)
    viewer:switchToImageNum(math.min(math.max(1, page_num), count))
end
```

Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua]: Now that's neat! I haven't seen that before, I'll implement it in the next commit.

File: plugins/opds.koplugin/opdsbrowser.lua [hunk 7]
```diff
@@ -804,12 +856,16 @@ function OPDSBrowser:showDownloads(item)
...
-    if (#type_buttons % 2 == 1) then -- we need even number of type buttons
-        table.insert(type_buttons, {text = ""})
-    end
-    for i = 2, #type_buttons, 2 do
-        table.insert(buttons, {type_buttons[i - 1], type_buttons[i]}) -- type buttons, two in a row
+    -- handles adding download button(s)
+    if (download_buttons ~= {}) then
+        if (#download_buttons == 1) then
+            table.insert(buttons, download_buttons)
+            table.insert(buttons, {})  -- add separator for stream buttons
```
Comment (hius07) [inline plugins/opds.koplugin/opdsbrowser.lua:863]: Do you need a separator when there are more than 1 download button?

File: plugins/opds.koplugin/opdsbrowser.lua [hunk 8]
```diff
@@ -804,12 +856,16 @@ function OPDSBrowser:showDownloads(item)
...
+        elseif (#download_buttons % 2 == 1) then
+            table.insert(download_buttons, {text = ""})
+        end
     end
+    table.insert(buttons, stream_buttons)
```
Comment (hius07) [inline plugins/opds.koplugin/opdsbrowser.lua:868]: It will produce additional separator when there are no stream buttons.
You can declare `local stream_buttons` and then initialize it as an array of two buttons if the stream exists.

Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua:868]: What If i just implement a check to see if any stream buttons exist?
```
    if (stream_buttons ~= {}) then
        table.insert(buttons, stream_buttons)
        table.insert(buttons, {}) -- separator
    end
```

File: plugins/opds.koplugin/opdsbrowser.lua (hunk 7 above)
Comment (hius07) [reply plugins/opds.koplugin/opdsbrowser.lua:863]: Okay, what if we have one download button and no stream?

Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua:863]: I left a comment about what I could do on the other review you left.

File: plugins/opds.koplugin/opdsbrowser.lua (hunk 8 above)
Comment (hius07) [reply plugins/opds.koplugin/opdsbrowser.lua]: > * download buttons, but no stream buttons

There will be no separator for 2 download buttons case.

Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua]: I Gotcha, now the separator is after the logic that determines if there are more than 1 download button. Now, if there is any number of download buttons there will be a separator.

Comment (hius07) [reply plugins/opds.koplugin/opdsbrowser.lua]: It seems you've missed to insert download buttons (more than 1 case).

Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua]: The _One_ Time I feel confident in a solution!  :)

I'll fix it.

Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua]: yeah, so I read `table.insert(download_buttons, {text = ""})` as `table.insert(buttons, download_buttons)`, which is why I forgot to add `table.insert(buttons, download_buttons)` in the more than 1 case.

File: plugins/opds.koplugin/opdsbrowser.lua [hunk 9]
```diff
@@ -804,14 +856,20 @@ function OPDSBrowser:showDownloads(item)
...
     end
-    if (#type_buttons % 2 == 1) then -- we need even number of type buttons
-        table.insert(type_buttons, {text = ""})
+    -- handles adding download button(s)
+    if (download_buttons ~= {}) then
+        if (#download_buttons == 1) then
+            table.insert(buttons, download_buttons)
+        elseif (#download_buttons % 2 == 1) then
+            table.insert(download_buttons, {text = ""})
+            table.insert(buttons, download_buttons)
```
Comment (hius07) [inline plugins/opds.koplugin/opdsbrowser.lua:865]: Since you are inserting unconditionally, move it out of the "if".

Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua]: Done.

**Comment by hius07** [comment]
Looks good (not tested).

File: plugins/opds.koplugin/opdsbrowser.lua [hunk 10]
```diff
@@ -695,7 +695,44 @@ function OPDSBrowser:downloadFile(item, filename, remote_url)
...
 end
 
-function OPDSBrowser:streamPages(item, remote_url, count)
+-- This function shows a dialog with input fields
+-- for entering information for Page streaming from.
+-- an input page number
+function OPDSBrowser:jumpToPage(viewer, count)
+    local input_dialog
+    input_dialog = InputDialog:new{
+        title = "Enter Page Number",
```
Comment (Frenzie) [inline plugins/opds.koplugin/opdsbrowser.lua:704]:
```suggestion
        title = _("Enter page number"),
```

The _() is for translation, the string itself so it meets our style guide and will simply take the same translation as the other however many times it occurs.

File: plugins/opds.koplugin/opdsbrowser.lua [hunk 11]
```diff
@@ -695,7 +695,44 @@ function OPDSBrowser:downloadFile(item, filename, remote_url)
...
+        input = "",
+        input_type = "number",
+        input_hint = "Page # (1 - "..count..")",
```
Comment (Frenzie) [inline plugins/opds.koplugin/opdsbrowser.lua:707]: Use this one instead so it's greppable:
https://github.com/koreader/koreader/blob/ee35257f997ec3cc719d23212428d39b4173129d/frontend/ui/widget/keyvaluepage.lua#L390-L392

File: plugins/opds.koplugin/opdsbrowser.lua (hunk 10 above)
Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua:704]: I figured it was there for a reason, just didn't know exactly why. That's why I tried copying other code that was already in the opdsbrowser.lua file. I'll re-add it.

File: plugins/opds.koplugin/opdsbrowser.lua [hunk 12]
```diff
@@ -794,7 +842,11 @@ function OPDSBrowser:showDownloads(item)
             end
             if filetype then -- supported file type
                 local text = acquisition.title and acquisition.title or string.upper(filetype)
-                table.insert(type_buttons, {
+
+                -- Got the url decode from here, https://bit.ly/3F0h6sO
```
Comment (Frenzie) [inline plugins/opds.koplugin/opdsbrowser.lua:846]: No need for URL shorteners
```suggestion
                -- Got the url decode from here, https://help.interfaceware.com/code/details/urlcode-lua
```

But we actually have the same principle already:
https://github.com/koreader/koreader/blob/ee35257f997ec3cc719d23212428d39b4173129d/reader.lua#L95-L109

File: plugins/opds.koplugin/opdsbrowser.lua (hunk 10 above)
Comment (Frenzie) [reply plugins/opds.koplugin/opdsbrowser.lua:704]: You can just click commit suggestion here in GitHub and then pull locally.

File: plugins/opds.koplugin/opdsbrowser.lua (hunk 12 above)
Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua:846]: I Tried this function and it didn't work, is there any particular reason we should use this function over what's implemented currently?

Also, I will change the URL to it's unshortened version.

File: plugins/opds.koplugin/opdsbrowser.lua (hunk 11 above)
Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua:707]: Is this what you're looking for? `input_hint = "Page # ".."(" .. "1 - " .. count .. ")",`

File: plugins/opds.koplugin/opdsbrowser.lua (hunk 4 above)
Comment (Frenzie) [reply plugins/opds.koplugin/opdsbrowser.lua:704]: GitHub may call it a suggestion, but this is not a suggestion. ;-)
```suggestion
        title = _("Enter page number"),
```

File: plugins/opds.koplugin/opdsbrowser.lua (hunk 12 above)
Comment (Frenzie) [reply plugins/opds.koplugin/opdsbrowser.lua]: That was mostly just an aside, though generically speaking the answer would be maintainability. Doing the same thing in different ways makes it harder to work with the codebase. But what exactly do you mean by didn't work? If you mean the obvious difference with the + then there's a chance that's a bug. If you mean the file:// stuff that's not really relevant here. :-)

Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua]: Ah, so that's my bad :) I just copied the whole function into opdsbrowser to see if it works, but I must have either used the function wrong or misunderstood the components. Let me try again, and see if I can make it work.

File: plugins/opds.koplugin/opdsbrowser.lua (hunk 11 above)
Comment (Frenzie) [reply plugins/opds.koplugin/opdsbrowser.lua]: Apologies for being unclear. My objection is to the "Page #". Forget about the rest.

**Review by NiLuJe — COMMENTED** [review]
Goddammit, I hate GH's review UI.

(I'd forgotten to hit submit on this).

File: plugins/opds.koplugin/opdsbrowser.lua [hunk 13]
```diff
@@ -794,7 +839,11 @@ function OPDSBrowser:showDownloads(item)
...
+                text = string.gsub(text, "+", " ")
+                text = string.gsub(text, "%%(%x%x)", function(x) return string.char(tonumber(x,16)) end)
```
[the remark this answers is not in the document]
Comment (Frenzie) [reply plugins/opds.koplugin/opdsbrowser.lua]: > LuaSocket provides url.unescape to do this in a slightly neater way ;).

I figured it didn't because it wasn't used here — d'oh. :laughing:

[the remark this answers is not in the document]
Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua]: > (And, FYI, actual upstream of this ref is https://github.com/lunarmodules/cgilua/blob/master/src/cgilua/urlcode.lua)

Thanks, I had a hard time finding the actual repo from where it was. Doesn't Matter anymore, though, i've removed that and replaced it with a different method. Would it be better to use the method @Frenzie mentioned, or LuaSocket method?

[the remark this answers is not in the document]
Comment (NiLuJe) [reply plugins/opds.koplugin/opdsbrowser.lua]: LuaSocket, definitely ;).

[the remark this answers is not in the document]
Comment (Frenzie) [reply plugins/opds.koplugin/opdsbrowser.lua]: LuaSocket here, dealing with URLs and protocols and such is what it's already there for in the file after all.

[the remark this answers is not in the document]
Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua]: Understood! Looks like I messed up the last commit anyway...

[the remark this answers is not in the document]
Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua]: Should be in the newest commit.

**Comment by bigdale123** [comment]
Just found that separators aren't working correctly anymore. I'll fix it.

File: plugins/opds.koplugin/opdsbrowser.lua [hunk 14]
```diff
@@ -804,14 +854,19 @@ function OPDSBrowser:showDownloads(item)
             end
         end
     end
-    if (#type_buttons % 2 == 1) then -- we need even number of type buttons
-        table.insert(type_buttons, {text = ""})
+    -- handles adding download button(s)
+    if (download_buttons ~= {} and #download_buttons > 0) then
```
Comment (hius07) [inline plugins/opds.koplugin/opdsbrowser.lua:858]: Isn't it the same?

Also please locate download buttons by two in a row, as it was.

Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua:858]: The way it was before did not work, the separator would still appear if download_buttons was empty. The >0 check fixes that issue. 

As for the second thing, how do you mean "as it was"?

Comment (hius07) [reply plugins/opds.koplugin/opdsbrowser.lua:858]: https://github.com/koreader/koreader/blob/ee35257f997ec3cc719d23212428d39b4173129d/plugins/opds.koplugin/opdsbrowser.lua#L810-L812

Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua:858]: Oh, I forgot to put that back in. Thanks!

**Comment by bigdale123** [comment]
Anything else I need to fix?

File: plugins/opds.koplugin/opdsbrowser.lua (hunk 4 above)
Comment (Frenzie) [reply plugins/opds.koplugin/opdsbrowser.lua:704]: This one still needs to be fixed.

File: plugins/opds.koplugin/opdsbrowser.lua [hunk 15]
```diff
@@ -695,7 +695,44 @@ function OPDSBrowser:downloadFile(item, filename, remote_url)
...
+-- This function shows a dialog with input fields
+-- for entering information for Page streaming from.
+-- an input page number
+function OPDSBrowser:jumpToPage(viewer, count)
+    local input_dialog
+    input_dialog = InputDialog:new{
+        title = _("Enter Page Number"),
+        input = "",
+        input_type = "number",
+        input_hint = "Page # ".."(" .. "1 - " .. count .. ")",
```
Comment (Frenzie) [inline plugins/opds.koplugin/opdsbrowser.lua:707]:
```suggestion
        input_hint = "(" .. "1 - " .. count .. ")",
```

Comment (bigdale123) [reply plugins/opds.koplugin/opdsbrowser.lua:707]: Oh, that's what you meant. Sorry!

File: plugins/opds.koplugin/opdsbrowser.lua (hunk 4 above)
Comment (Frenzie) [reply plugins/opds.koplugin/opdsbrowser.lua:704]: And `Enter page number`, not `Enter Page Number`. ;-)

File: plugins/opds.koplugin/opdsbrowser.lua (hunk 15 above)
Comment (Frenzie) [reply plugins/opds.koplugin/opdsbrowser.lua]: As written it would've put the untranslated string "Page #" on the screen, but it's not really necessary to put "Page number/#" everywhere you can regardless. So I think this way it's a win/win.

For future reference, the correct way to write this if we wanted it localized would be:
```lua
T(_("Page # (1 - %1)"), count)
```
But like I said, I don't want to see Page # here regardless. ^_^

For more info see:
http://koreader.rocks/doc/modules/ffi.util.html#template
http://koreader.rocks/doc/modules/gettext.html

File: plugins/opds.koplugin/opdsbrowser.lua [hunk 16]
```diff
@@ -904,6 +966,8 @@ function OPDSBrowser:browse(browse_url, username, password)
     end
 end
 
+
+
```
Comment (Frenzie) [inline plugins/opds.koplugin/opdsbrowser.lua:970]: Some stray whitespace snuck in.
```suggestion
```

**Review by Frenzie — APPROVED** [review]
lgtm besides the whitespace nit.

**Comment by bigdale123** [comment]
> lgtm besides the whitespace nit.

Where is the whitespace nit issue? I can fix it.

**Comment by Frenzie** [comment]
Line 969 and 970 were added.

**Comment by bigdale123** [comment]
Should be fixed.

**Comment by Frenzie** [comment]
There seems to be some kind of merge conflict, though it's not quite clear to me what it might be from GH alone.

**Comment by bigdale123** [comment]
Let me know if there is anything I can do on my side to help resolve the conflict.

**Comment by Frenzie** [comment]
It looks like this is your master branch rather than a separate feature branch, and that it includes some things relating to the namespace. That's probably the cause of the conflict. What did you guys do to resolve that last week, just open a new PR from a new branch with the updated file?

**Comment by bigdale123** [comment]
> It looks like this is your master branch rather than a separate feature branch, and that it includes some things relating to the namespace. That's probably the cause of the conflict. What did you guys do to resolve that last week, just open a new PR from a new branch with the updated file?

I'm not 100% Sure, but I will remember to put new work on a different branch next time.

**Comment by bigdale123** [comment]
I tried going back and looking at the last PR I made, and I still don't know how it got merged without conflict last time. At this point, should I just close this and make a new PR with the changes in a different branch?

**Comment by Frenzie** [comment]
I don't have the time to investigate in Git atm so from my end the answer's affirmative.

**Comment by bigdale123** [comment]
I found the issue... My fork was out of date, is there still a merge conflict after fetching and merging the new commits?

**Comment by Frenzie** [comment]
Yes, it conflicts because it starts with the OPDS namespace stuff, although it's not really clear to me why GH can show a sane looking diff but not simply merge that diff.

**Comment by poire-z** [comment]
Github didn't show me any "conflict":
image

**Comment by bigdale123** [comment]
AUGH. Too late now... :), i'll finish the new PR in a second.

**Comment by bigdale123** [comment]
Done, sorry again for all the trouble.

**Comment by Frenzie** [comment]
All's well that ends well given the object comparison issue. :-)
